lte/gateway/c/core/oai/test/mme_app_task/test_mme_procedures.cpp

```diff
@@ -0,0 +1,372 @@
+/**
+ * Copyright 2020 The Magma Authors.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <chrono>
+#include <gtest/gtest.h>
+#include <thread>
+
+#include "feg/protos/s6a_proxy.pb.h"
+#include "../mock_tasks/mock_tasks.h"
+#include "mme_app_state_manager.h"
+#include "mme_app_ip_imsi.h"
+#include "proto_msg_to_itti_msg.h"
+
+extern "C" {
+#include "common_types.h"
+#include "mme_config.h"
+#include "mme_app_defs.h"
+#include "mme_app_extern.h"
+#include "mme_app_state.h"
+#include "security_types.h"
+}
+
+task_zmq_ctx_t task_zmq_ctx_main;
+
+static int handle_message(zloop_t* loop, zsock_t* reader, void* arg) {
+  MessageDef* received_message_p = receive_msg(reader);
+
+  switch (ITTI_MSG_ID(received_message_p)) {
+    default: {
+    } break;
+  }
+
+  itti_free_msg_content(received_message_p);
+  free(received_message_p);
+  return 0;
+}
+static void send_mme_app_uplink_data_ind(
+    uint8_t* nas_msg, uint8_t nas_msg_length, const plmn_t& plmn) {
+  MessageDef* message_p =
+      itti_alloc_new_message(TASK_S1AP, MME_APP_UPLINK_DATA_IND);
+  ITTI_MSG_LASTHOP_LATENCY(message_p)     = 0;
+  MME_APP_UL_DATA_IND(message_p).ue_id    = 1;
+  MME_APP_UL_DATA_IND(message_p).nas_msg  = blk2bstr(nas_msg, nas_msg_length);
+  MME_APP_UL_DATA_IND(message_p).tai.plmn = plmn;
+  MME_APP_UL_DATA_IND(message_p).tai.tac  = 1;
+  MME_APP_UL_DATA_IND(message_p).cgi.plmn = plmn;
+  MME_APP_UL_DATA_IND(message_p).cgi.cell_identity = {0, 0, 0};
+  send_msg_to_task(&task_zmq_ctx_main, TASK_MME_APP, message_p);
+  return;
+}
+
+class MmeAppProcedureTest : public ::testing::Test {
+  virtual void SetUp() {
+    s1ap_handler = std::make_shared<MockS1apHandler>();
+    s6a_handler  = std::make_shared<MockS6aHandler>();
+    spgw_handler = std::make_shared<MockSpgwHandler>();
+
+    itti_init(
+        TASK_MAX, THREAD_MAX, MESSAGES_ID_MAX, tasks_info, messages_info, NULL,
+        NULL);
+
+    // initialize mme config
+    mme_config_init(&mme_config);
+    create_partial_lists(&mme_config);
+    mme_config.use_stateless                              = false;
+    mme_config.nas_config.prefered_integrity_algorithm[0] = EIA2_128_ALG_ID;
+
+    task_id_t task_id_list[10] = {
+        TASK_MME_APP,    TASK_HA,  TASK_S1AP,   TASK_S6A,      TASK_S11,
+        TASK_SERVICE303, TASK_SGS, TASK_SGW_S8, TASK_SPGW_APP, TASK_SMS_ORC8R};
+    init_task_context(
+        TASK_MAIN, task_id_list, 10, handle_message, &task_zmq_ctx_main);
```

Comment:
Might be a good candidate for a common setup function later.

Reply:
as we add more tests, it will be more clear how to refactor.